storage/mutable_segment.go

```diff
@@ -159,12 +160,87 @@ func (s *mutableSeg) Intersects(startNanosInclusive, endNanosExclusive int64) bo
 	return intersects
 }
 
-// TODO(xichen): Implement this.
 func (s *mutableSeg) QueryRaw(
 	ctx context.Context,
 	q query.ParsedRawQuery,
 ) ([]query.RawResult, error) {
-	return nil, errors.New("not implemented")
+	// Fast path if the limit indicates no results are needed.
+	if q.Limit <= 0 {
+		return nil, nil
+	}
+
+	// Retrieve the fields.
+	s.RLock()
+	if s.closed {
+		s.RUnlock()
+		return nil, errMutableSegmentAlreadyClosed
+	}
+	if s.sealed {
+		s.RUnlock()
+		return nil, errMutableSegmentAlreadySealed
+	}
+
+	allowedFieldTypes, fieldIndexMap, queryFields, err := s.collectFieldsForRawQueryWithLock(q)
+	if err != nil {
+		s.RUnlock()
+		return nil, err
+	}
+	defer func() {
+		for i := range queryFields {
+			if queryFields[i] != nil {
+				queryFields[i].Close()
+				queryFields[i] = nil
+			}
+		}
+	}()
+
+	numDocuments := s.mutableSegmentBase.NumDocuments()
+	s.RUnlock()
```

Comment:
just curious, if we want to give up the mutex ASAP, can we move the defer block from L188 after the mutex unlock? Will that actually make any difference?

Reply:
Probably no difference but doesn't hurt either, can move.